modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/HttpHeadersValidator.java

```diff
@@ -0,0 +1,168 @@
+        return new Netty4HttpHeaderValidator((httpRequest, channel, listener) -> {
+            if (httpRequest.headers() instanceof ValidatableHttpHeaders validatableHttpHeaders) {
+                // make sure validation only runs on properly wrapped "validatable" headers implementation
+                validator.apply(asHttpPreRequest(httpRequest), channel, ActionListener.wrap(validationResult -> {
+                    validatableHttpHeaders.markAsSuccessfullyValidated(validationResult);
+                    // a successful validation needs to signal to the {@link Netty4HttpHeaderValidator} to resume
+                    // forwarding the request beyond the headers part
+                    listener.onResponse(null);
+                }, e -> listener.onFailure(new HttpHeadersValidationException(e))));
+            } else {
+                listener.onFailure(new IllegalStateException("Expected a validatable request for validation"));
+            }
+        });
+    }
+
+    /**
+     * Given a {@link DefaultHttpRequest} argument, this returns a new {@link DefaultHttpRequest} instance that's identical to the
+     * passed-in one, but the headers of the latter are "validatable", in the sense that the channel handlers returned by
+     * {@link #getValidatorInboundHandler()} can use to convey the validation result.
+     */
+    public static HttpMessage wrapAsValidatableMessage(HttpMessage newlyDecodedMessage) {
+        assert newlyDecodedMessage instanceof HttpRequest;
+        DefaultHttpRequest httpRequest = (DefaultHttpRequest) newlyDecodedMessage;
+        ValidatableHttpHeaders validatableHttpHeaders = new ValidatableHttpHeaders(newlyDecodedMessage.headers());
+        return new DefaultHttpRequest(httpRequest.protocolVersion(), httpRequest.method(), httpRequest.uri(), validatableHttpHeaders);
+    }
+
+    public static ValidationResult extractValidationResult(org.elasticsearch.http.HttpRequest request) {
+        ValidatableHttpHeaders authenticatedHeaders = unwrapValidatableHeaders(request);
+        return authenticatedHeaders != null ? authenticatedHeaders.validationResultContextSetOnce.get() : null;
+    }
+
+    private static ValidatableHttpHeaders unwrapValidatableHeaders(org.elasticsearch.http.HttpRequest request) {
+        if (request instanceof Netty4HttpRequest == false) {
+            return null;
+        }
+        if (((Netty4HttpRequest) request).getNettyRequest().headers() instanceof ValidatableHttpHeaders == false) {
+            return null;
+        }
+        return (ValidatableHttpHeaders) (((Netty4HttpRequest) request).getNettyRequest().headers());
+    }
+
+    /**
+     * {@link HttpHeaders} implementation that carries along the {@link ValidationResult} iff
+     * the HTTP headers have been validated successfully.
+     */
+    public static final class ValidatableHttpHeaders extends DefaultHttpHeaders {
+
+        @FunctionalInterface
+        public interface ValidationResult {
```

Comment:
I don't think we need to skirt around that this is the result of authentication or try to hide that this is actually a ThreadContext.StoredContext by using different name.

Reply:
Okay, I've pushed [edc2a1d](https://github.com/elastic/elasticsearch/pull/95112/commits/edc2a1d20298cb22d5d8c167e7bc6ee5055823f7)